### PR TITLE
add support for `sku_name` and `trace` for `azurerm_spring_cloud_service`

### PR DESCRIPTION
**Pull request by njuCZ**
image

### CODE REVIEW

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 1]
```diff
@@ -133,8 +144,38 @@ func resourceArmSpringCloudService() *schema.Resource {
 				},
 			},
 
+			"trace": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:150]: this doesn't want to be computed:

```suggestion
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 2]
```diff
@@ -133,8 +144,38 @@ func resourceArmSpringCloudService() *schema.Resource {
...
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:156]: rather than having `enabled` as a field here - we can use the presence of this block to assume `enabled = true` and the ommission of this block to mean `enabled=false` - so we can remove this field

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 3]
```diff
@@ -133,8 +144,38 @@ func resourceArmSpringCloudService() *schema.Resource {
...
+						},
+					},
+				},
+			},
+
 			"tags": tags.Schema(),
 		},
+
+		// "sku_name" could update from "B0" to "S0"
+		CustomizeDiff: customdiff.ForceNewIfChange("sku_name", func(old, new, meta interface{}) bool {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:171]: it's worth noting this won't work if this field is Interpolated, which it likely will be - so unfortunately this won't catch all users

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 4]
```diff
@@ -422,6 +480,17 @@ func expandArmSpringCloudGitPatternRepository(input []interface{}) (*[]appplatfo
 	return &results, nil
 }
 
+func expandArmSpringCloudTrace(input []interface{}) *appplatform.TraceProperties {
+	if len(input) == 0 || input[0] == nil {
+		return nil
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:485]: (as above) we can take the omission of this block to mean this is Disabled - so we can make this:

```suggestion
		return &appplatform.TraceProperties{
			Enabled: utils.Bool(false),
		}
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 5]
```diff
@@ -618,3 +687,25 @@ func flattenArmSpringCloudGitPatternRepository(input *[]appplatform.GitPatternRe
...
+	enabled := false
+	appInsightInstrumentationKey := ""
+	if input.Enabled != nil {
+		enabled = *input.Enabled
+	}
+	if input.AppInsightInstrumentationKey != nil {
+		appInsightInstrumentationKey = *input.AppInsightInstrumentationKey
+	}
+
+	return []interface{}{
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:705]: if this is Disabled we can return an empty list here:

```suggestion
	if !enabled {
		return []interface{}{}
	}
```

File: azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go [hunk 6]
```diff
@@ -186,6 +213,27 @@ resource "azurerm_spring_cloud_service" "test" {
...
+  location = "%s"
+}
+
+resource "azurerm_spring_cloud_service" "test" {
+  name                = "acctest-sc-%d"
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+  sku_name            = "B0"
+}
+
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go:233]: I think this'll cause terrafmt to complain:

```suggestion
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 7]
```diff
@@ -618,3 +687,25 @@ func flattenArmSpringCloudGitPatternRepository(input *[]appplatform.GitPatternRe
...
+		map[string]interface{}{
+			"enabled":                         enabled,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:707]: which means we can remove this:

```suggestion
```

File: website/docs/r/spring_cloud_service.html.markdown [hunk 8]
```diff
@@ -109,6 +126,14 @@ The `ssh_auth` block supports the following:
 
 * `strict_host_key_checking_enabled` - (Optional) Indicates whether the Config Server instance will fail to start if the host_key does not match.
 
+---
+
+The `trace` block supports the following:
+
+* `enabled` - (Required) Indicates whether the Trace is enabled.
+
+* `app_insight_instrumentation_key` - (Optional) The instrumentationKey of Application Insights.
```
Comment (tombuildsstuff) [inline website/docs/r/spring_cloud_service.html.markdown:135]: can we make this:

```suggestion
* `instrumentation_key` - (Optional) The Instrumentation Key used for Application Insights.
```

File: website/docs/r/spring_cloud_service.html.markdown [hunk 9]
```diff
@@ -22,17 +22,30 @@ resource "azurerm_resource_group" "example" {
...
 
   config_server_git_setting {
     uri          = "https://github.com/Azure-Samples/piggymetrics"
     label        = "config"
     search_paths = ["dir1", "dir2"]
   }
 
+  trace {
+    enabled                         = true
+    app_insight_instrumentation_key = azurerm_application_insights.example.instrumentation_key
```
Comment (tombuildsstuff) [inline website/docs/r/spring_cloud_service.html.markdown:46]: we can shorten this down to:

```suggestion
    instrumentation_key = azurerm_application_insights.example.instrumentation_key
```

File: website/docs/r/spring_cloud_service.html.markdown [hunk 10]
```diff
@@ -22,17 +22,30 @@ resource "azurerm_resource_group" "example" {
...
+  sku_name            = "S0"
 
   config_server_git_setting {
     uri          = "https://github.com/Azure-Samples/piggymetrics"
     label        = "config"
     search_paths = ["dir1", "dir2"]
   }
 
+  trace {
+    enabled                         = true
```
Comment (tombuildsstuff) [inline website/docs/r/spring_cloud_service.html.markdown:45]: (as above) if we use the presence of the block to infer this we can make this:

```suggestion
```

File: azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go [hunk 11]
```diff
@@ -233,12 +288,17 @@ resource "azurerm_spring_cloud_service" "test" {
     }
   }
 
+  trace {
+    enabled                         = true
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go:292]: (as above) if we use the presence of the block to infer this we can remove this:

```suggestion
```

File: azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go [hunk 12]
```diff
@@ -233,12 +288,17 @@ resource "azurerm_spring_cloud_service" "test" {
...
+    app_insight_instrumentation_key = azurerm_application_insights.test.instrumentation_key
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go:293]: we can simplify this down to:

```suggestion
    instrumentation_key = azurerm_application_insights.test.instrumentation_key
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 13]
```diff
@@ -618,3 +687,25 @@ func flattenArmSpringCloudGitPatternRepository(input *[]appplatform.GitPatternRe
...
+			"app_insight_instrumentation_key": appInsightInstrumentationKey,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:708]: (as above) we can simplify this to:

```suggestion
			"instrumentation_key": appInsightInstrumentationKey,
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 14]
```diff
@@ -422,6 +480,17 @@ func expandArmSpringCloudGitPatternRepository(input []interface{}) (*[]appplatfo
...
 }
 
+func expandArmSpringCloudTrace(input []interface{}) *appplatform.TraceProperties {
+	if len(input) == 0 || input[0] == nil {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+	return &appplatform.TraceProperties{
+		Enabled:                      utils.Bool(v["enabled"].(bool)),
+		AppInsightInstrumentationKey: utils.String(v["app_insight_instrumentation_key"].(string)),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:490]:
```suggestion
		AppInsightInstrumentationKey: utils.String(v["instrumentation_key"].(string)),
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 15]
```diff
@@ -422,6 +480,17 @@ func expandArmSpringCloudGitPatternRepository(input []interface{}) (*[]appplatfo
 	return &results, nil
 }
 
+func expandArmSpringCloudTrace(input []interface{}) *appplatform.TraceProperties {
+	if len(input) == 0 || input[0] == nil {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+	return &appplatform.TraceProperties{
+		Enabled:                      utils.Bool(v["enabled"].(bool)),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:489]: since the presence of this block infers this is enabled - we can hard-code this:

```suggestion
		Enabled:                      utils.Bool(true),
```

File: azurerm/internal/services/appplatform/spring_cloud_service_resource.go [hunk 16]
```diff
@@ -133,8 +144,38 @@ func resourceArmSpringCloudService() *schema.Resource {
...
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+
+						"app_insight_instrumentation_key": {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/appplatform/spring_cloud_service_resource.go:159]: we can shorten this to `instrumentation_key` - since no other Azure Service uses this terminology

**Comment by njuCZ** [comment]
@tombuildsstuff thanks for your kindly suggestion. I have modified this PR and passed the acctest, could you please have a look again?
It seems the CI failed error message is not related with this PR:
```
Error response from daemon: No such container: tf-website-azurerm-temp
Makefile:85: recipe for target 'website-provider-test' failed
```

**Comment by tombuildsstuff** [comment]
@njuCZ I've re-triggered Travis so that should go green, looks like a connection error

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2020-07-01 at 17 46 43
